src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala

```diff
@@ -34,100 +38,165 @@ object StructuralVariant {
+    def buildVariantGraph(exceptionalReads: Iterable[PairedMappedRead], maxNormalInsertSize: Int): PairedReadGraph = {
+      val reads = exceptionalReads.toArray.sortBy(r => Math.min(r.read.start, r.mate.start))
+      val graph = MutableGraph[PairedMappedRead, WUnDiEdge]()
+
+      for {
+        i <- 0 until reads.length
+      } {
+        val read = reads(i)
+        val start = Math.min(read.read.start, read.mate.start)
+        var j = i + 1
+        var done = false
+        // Loop through subsequent reads until it's impossible to find another compatible one.
+        while (j < reads.length && !done) {
+          val nextRead = reads(j)
+          val nextStart = Math.min(nextRead.read.start, nextRead.mate.start)
+          if (Math.abs(nextStart + nextRead.read.sequence.length - start) > maxNormalInsertSize) {
+            done = true
+          } else {
+            if (!areReadsIncompatible(read, nextRead, maxNormalInsertSize)) {
+              graph += (read ~ nextRead) % 1
+            }
+          }
+          j += 1
+        }
       }
+
+      // TODO: exclude isolated reads from the graph
+      graph
     }
 
     override def run(args: Arguments, sc: SparkContext): Unit = {
       val readSet = Common.loadReadsFromArguments(args, sc, Read.InputFilters(nonDuplicate = true))
       val pairedMappedReads = readSet.mappedPairedReads
-      val firstInPair = pairedMappedReads.filter(_.isFirstInPair)
-
-      // Pare down to mate pairs which aren't highly displaced & aren't inverted.
-      val readsInRange = for {
-        r <- firstInPair
-        s <- orientedInsertSize(r)
-        if 0 < s && s < MAX_INSERT_SIZE
-      } yield r
-      readsInRange.persist()
+      val firstInPair = pairedMappedReads.filter(_.isFirstInPair).flatMap(PairedMappedRead(_))
 
-      val insertSizes = readsInRange.flatMap(orientedInsertSize)
-      val insertStats = insertSizes.stats()
-      println("Stats on inferredInsertSize: " + insertStats)
-
-      // Find blocks where the average insert size is 2 sigma from the average for the whole genome.
-      val exceptionalThreshold = insertStats.mean + 2 * insertStats.stdev
-      val exceptionalBlocks = (for {
-        r <- readsInRange
-        pos <- matedReadBlocks(r)
-        insertSize <- orientedInsertSize(r)
-      } yield {
-        Location(r.read.referenceContig, pos) -> insertSize
-      })
-        .aggregateByKey(new StatCounter())(seqOp = _.merge(_), combOp = _.merge(_))
-        .filter { case (loc, stats) => stats.mean > exceptionalThreshold }
-        .map(_._1)
-
-      // Group adjacent blocks on the same contig
-      val exceptionalRanges =
-        for {
-          (contig, locations) <- exceptionalBlocks.groupBy(_.contig)
-          positions = locations.map(_.position).toArray.sorted
-          (start, end) <- coalesceAdjacent(positions, BLOCK_SIZE)
-        } yield {
-          GenomeRange(contig, start, end)
-        }
+      val packed = getExceptionalReads(firstInPair)
```

Comment:
possibly:
```
val ExceptionalReadsReturnType(_, _, _, maxNormalInsertSize, exceptionalReads) = getExceptionalReads(firstInPair)
```

rather than these three lines